patches/tModLoader/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs.patch

```diff
@@ -0,0 +1,219 @@
++			vanillaLayers.Add(PlayerDrawLayer.Pulley);
++			vanillaLayers.Add(PlayerDrawLayer.Shield);
++			vanillaLayers.Add(PlayerDrawLayer.MountIndent); //{
++			vanillaLayers.Add(PlayerDrawLayer.SolarShield);
++			vanillaLayers.Add(PlayerDrawLayer.MountUnindent); //}
+ 			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindFrontArm)
+-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
++				vanillaLayers.Add(PlayerDrawLayer.HeldItem);
+ 
+-			PlayerDrawLayers.DrawPlayer_28_ArmOverItem(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_29_OnhandAcc(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_30_BladedGlove(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
++			vanillaLayers.Add(PlayerDrawLayer.ArmOverItem);
++			vanillaLayers.Add(PlayerDrawLayer.HandOnAcc);
++			vanillaLayers.Add(PlayerDrawLayer.BladedGlove);
++			vanillaLayers.Add(PlayerDrawLayer.TorsoUnindent); //}
+ 			if (!drawinfo.drawFrontAccInNeckAccLayer)
+-				PlayerDrawLayers.DrawPlayer_32_FrontAcc(ref drawinfo);
++				vanillaLayers.Add(PlayerDrawLayer.FrontAcc);
+ 
+ 			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.OverFrontArm)
+-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
+-
+-			PlayerDrawLayers.DrawPlayer_31_ProjectileOverArm(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_33_FrozenOrWebbedDebuff(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_34_ElectrifiedDebuffFront(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_35_IceBarrier(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_36_CTG(ref drawinfo);
+-			PlayerDrawLayers.DrawPlayer_37_BeetleBuff(ref drawinfo);
++				vanillaLayers.Add(PlayerDrawLayer.HeldItem);
++
++			vanillaLayers.Add(PlayerDrawLayer.ProjectileOverArm);
++			vanillaLayers.Add(PlayerDrawLayer.FrozenOrWebbedDebuff);
++			vanillaLayers.Add(PlayerDrawLayer.ElectrifiedDebuffFront);
++			vanillaLayers.Add(PlayerDrawLayer.IceBarrier);
++			vanillaLayers.Add(PlayerDrawLayer.CaptureTheGem);
++			vanillaLayers.Add(PlayerDrawLayer.BeetleBuff);
++
++			for (int i = 0; i < vanillaLayers.Count; i++) {
++				var layer = (LegacyPlayerDrawLayer)vanillaLayers[i];
++
++				layer.defaultDepth = layer.depth = i;
++			}
++
++			var layers = new Dictionary<string, List<PlayerDrawLayer>> {
++				{ "Terraria", vanillaLayers }
++			};
++
++			//Add OOP layers.
++			PlayerLayerHooks.AddDrawLayers(drawPlayer, layers);
++			//Add legacy layers.
++			PlayerHooks.AddDrawLayers(drawPlayer, layers);
++
++			//Actually make layer lists readonly
++			var readonlyLayersBackingDictionary = new Dictionary<string, IReadOnlyList<PlayerDrawLayer>>();
++			var readonlyLayers = new ReadOnlyDictionary<string, IReadOnlyList<PlayerDrawLayer>>(readonlyLayersBackingDictionary);
++
++			foreach (var pair in layers) {
++				readonlyLayersBackingDictionary[pair.Key] = pair.Value.AsReadOnly();
++			}
++
++			//Modify draw layers, but not the collections.
++			PlayerHooks.ModifyDrawLayers(drawPlayer, readonlyLayers);
```

Comment:
should there be a flag on a layer to disable it?

Reply:
Ye, all the collections are readonly, but visibility and position can still be modified there.